kubernetes/resource_kubernetes_default_service_account.go

```diff
@@ -59,6 +60,38 @@ func resourceKubernetesDefaultServiceAccountCreate(ctx context.Context, d *schem
 		return diag.FromErr(err)
 	}
 
+	ops := patchMetadata("metadata.0.", "/metadata/", d)
+	imagePullSecrets := d.Get("image_pull_secret").(*schema.Set).List()
+	ops = append(ops, &ReplaceOperation{
+		Path:  "/imagePullSecrets",
+		Value: expandLocalObjectReferenceArray(imagePullSecrets),
+	})
+
+	secrets := d.Get("secret").(*schema.Set).List()
+	defaultSecretName := d.Get("default_secret_name").(string)
+
```

Comment:
Thanks @arybolovlev

Since the default service account is already created by K8s `default_secret_name` shouldn't return an empty string, same as https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/kubernetes/resource_kubernetes_service_account.go#L319

Please correct me if I'm wrong.

Reply:
@arybolovlev it would be great if you can re-review.

Reply:
Hi @husunal,

`default_secret_name` is something that doesn't exist in the Kubernetes and is purely computed by this provider. It means that you cannot configure this attribute in your Terraform code. If you try, you will get the following error:
```
Can't configure a value for "default_secret_name": its value will be decided automatically based on the result of applying this configuration.
```
That happens on the schema validation step before any CRUD operations.

The create operation that you have added is called only when Terraform detects that the given code has a resource that doesn't exist in the state file. Then it gets resource attributes and does whatever is necessary with them to call Kubernetes API. At this point, since we cannot use attribute `default_secret_name` in our Terraform code, the `Get` method will return the empty string. The function, that you call later, `expandServiceAccountSecrets` can handle case when one of the arguments `defaultSecretName` is empty string. If we know that the string in our case in empty all the time, let just pass emtpy string instead of getting it.

Actually, I found another interesting behaviour here.
I have the following Terraform code:
```hcl
resource "kubernetes_default_service_account" "this" {
  metadata {
    namespace = "this"
  }

  automount_service_account_token = false
}
```
As you can see `image_pull_secrets` and `secrets` attributes are ommited. Below is a JSON that will be passed to `Patch` method:
```
[
    {"path":"/imagePullSecrets","value\":[],"op":"replace"},
    {"path":"/secrets","value":[],"op":"replace"},
    {"path":"/automountServiceAccountToken","value":false,"op":"replace"}
]
```
It means that `imagePullSecrets` and `secrets` will be erased from the default service account on the Kubernetes end if they don't set in Terraform code. Moreover, if you pass `secrets` and none of them has type `kubernetes.io/service-account-token`, then a new token will be generated for the default service account.

I would say, that we need to check if length of `imagePullSecrets` and `secrets` more than 0 and only then append `ops`. In this case, I would also recomment to replace `defaultSecretName := d.Get("default_secret_name").(string)` with function `getServiceAccountDefaultSecret` that you use below to get default secret name. Then we don't need to pass empty string to function `expandServiceAccountSecrets`.

Thank you.